biofefi/machine_learning/learner.py

```diff
@@ -39,21 +112,42 @@ def fit(self, data: Tuple) -> None:
             return res, metric_res, metric_res_stats, trained_models
 
     def _fit_holdout(self, data: Tuple) -> None:
+        """
+        Trains models using a holdout strategy with bootstrap sampling.
+
+        Args:
+            - data (Tuple): Tuple containing training and testing data
+            for each bootstrap sample.
+
+        Returns:
+            - res (Dict): Dictionary containing model predictions for
+            each bootstrap sample.
+            - metric_res (Dict): Dictionary containing metric values for
+            each bootstrap sample.
+            - metric_res_stats (Dict): Dictionary containing average and
+            standard deviation of metric values across bootstrap samples.
+            - trained_models (Dict): Dictionary containing
+            trained models for each model type.
+        """
         self._logger.info("Fitting holdout with bootstrapped datasets...")
         res = {}
         metric_res = {}
         trained_models = {model_name: [] for model_name in self._models.keys()}
 
         for i in range(self._n_bootstraps):
             self._logger.info(f"Processing bootstrap sample {i+1}...")
-            X_train, X_test = data.X_train[i], data.X_test[i]
-            y_train, y_test = data.y_train[i], data.y_test[i]
+            X_train, X_test, y_train, y_test = self._process_data_for_bootstrap(data, i)
 
             res[i] = {}
             for model_name, model in self._models.items():
                 res[i][model_name] = {}
                 self._logger.info(f"Fitting {model_name} for bootstrap sample {i+1}...")
-                model.fit(X_train, y_train)
+
+                if model_name in [ModelNames.BRNNClassifier, ModelNames.BRNNRegressor]:
+                    model.fit_model(X_train, y_train)
+                else:
+                    model.fit(X_train, y_train)
```

Comment:
This is why I asked you to make a wrapper class around the BRNN classes you made. I don't want to have the code checking if it's one of those models or not every time. The point of the `sklearn` wrappers is that our custom models work the some way as the `sklearn` models and we don't need to check. Otherwise, any time we stuff with any models, we need to check what the model is.

Take the two classes you have in `nn_models.py` and split them again into 2, so there's 4 classes.
- `BRNNClassifie`r: the `sklearn` compatible wrapper with the `fit` and `predict` methods that you need to override.
- `BayesianRegularisedNNClassifier`: your model whose training and predicting logic you'll fit into the wrapper
- `BRNNRegressor`: the `sklearn` compatible wrapper with the `fit` and `predict` methods that you need to override.
- `BayesianRegularisedNNRegressor`: your model whose training and predicting logic you'll fit into the wrapper

see this page for more info: https://scikit-learn.org/stable/developers/develop.html#rolling-your-own-estimator

Reply:
Hi Dan, 

Thanks for your feedback. I can still implement the expected behaviour by removing these codes and changing the method name from fit_model to fit method. 

`class BayesianRegularisedNNClassifier(BaseNetwork, BaseEstimator, ClassifierMixin):
pass

fit(self, X: np.ndarray, y: np.ndarray) -> None: 

instead of fit_model(self, X: np.ndarray, y: np.ndarray) -> None: 
`

`                
# I can remove these because it has a fit_model method for fitting and training. 

if model_name in [ModelNames.BRNNClassifier, ModelNames.BRNNRegressor]:
                    model.fit_model(X_train, y_train)
                else:
                    model.fit(X_train, y_train)
`
This approach allows us to remove the unnecessary code block while ensuring compatibility with the existing pipeline. Additionally, since I have implemented the ` predict`  method to align with the scikit-learn API, it seamlessly integrates and functions without errors, maintaining consistency across the pipeline.

Let me know if this aligns with your expectations, or if you'd still prefer wrapper classes for additional separation.

Reply:
I think that might work. Make sure the `fit` methods return `self` as that's what they appear to do in the documentation